components/style/values/specified/list.rs

```diff
@@ -9,8 +9,70 @@ use parser::{Parse, ParserContext};
 use std::fmt;
 use style_traits::{ParseError, StyleParseErrorKind, ToCss};
 use values::{Either, None_};
+#[cfg(feature = "gecko")]
+use values::CustomIdent;
+#[cfg(feature = "gecko")]
+use values::generics::CounterStyleOrNone;
 use values::specified::UrlOrNone;
 
+/// Specified and computed `list-style-type` property.
+#[cfg(feature = "gecko")]
+#[derive(Clone, Debug, Eq, MallocSizeOf, PartialEq, ToComputedValue, ToCss)]
+pub enum ListStyleType {
+    /// <counter-style> | none
+    CounterStyle(CounterStyleOrNone),
+    /// <string>
+    String(String),
+}
+
+#[cfg(feature = "gecko")]
+impl ListStyleType {
+    /// Initial specified value for `list-style-type`.
+    #[inline]
+    pub fn disc() -> Self {
+        ListStyleType::CounterStyle(CounterStyleOrNone::disc())
+    }
+
+    /// Convert from gecko keyword to list-style-type.
+    ///
+    /// This should only be used for mapping type attribute to
+    /// list-style-type, and thus only values possible in that
+    /// attribute is considered here.
+    pub fn from_gecko_keyword(value: u32) -> Self {
+        use gecko_bindings::structs;
+
+        let counter_style = if value == structs::NS_STYLE_LIST_STYLE_NONE {
+            CounterStyleOrNone::None
+        } else {
+            CounterStyleOrNone::Name(CustomIdent(match value {
+                structs::NS_STYLE_LIST_STYLE_DISC => atom!("disc"),
+                structs::NS_STYLE_LIST_STYLE_CIRCLE => atom!("circle"),
+                structs::NS_STYLE_LIST_STYLE_SQUARE => atom!("square"),
+                structs::NS_STYLE_LIST_STYLE_DECIMAL => atom!("decimal"),
+                structs::NS_STYLE_LIST_STYLE_LOWER_ROMAN => atom!("lower-roman"),
+                structs::NS_STYLE_LIST_STYLE_UPPER_ROMAN => atom!("upper-roman"),
+                structs::NS_STYLE_LIST_STYLE_LOWER_ALPHA => atom!("lower-alpha"),
+                structs::NS_STYLE_LIST_STYLE_UPPER_ALPHA => atom!("upper-alpha"),
+                _ => unreachable!("Unknown counter style keyword value"),
+            }))
+        };
+
+        ListStyleType::CounterStyle(counter_style)
+    }
+}
+
+#[cfg(feature = "gecko")]
+impl Parse for ListStyleType {
+    fn parse<'i, 't>(context: &ParserContext, input: &mut Parser<'i, 't>)
+                         -> Result<ListStyleType, ParseError<'i>> {
+        Ok(if let Ok(style) = input.try(|i| CounterStyleOrNone::parse(context, i)) {
+            ListStyleType::CounterStyle(style)
+        } else {
+            ListStyleType::String(input.expect_string()?.as_ref().to_owned())
+        })
```

Comment:
Will it be better to make this like following?

```rust
if let Ok(style) = input.try(|i| CounterStyleOrNone::parse(context, i)) {
    return Ok(ListStyleType::CounterStyle(style));
}

Ok(ListStyleType::String(input.expect_string()?.as_ref().to_owned()))
```

Besides, I'm wondering if using pattern matching would work?

↓ I'm not sure if this piece of code will work
```rust
match (input.try(|i| CounterStyleOrNone::parse(context, i))) {
  Ok(style) => Ok(ListStyleType::CounterStyle(style)),
  Err(_) => Ok(ListStyleType::String(input.expect_string()?.as_ref().to_owned()))
}
```

Reply:
Yeah I prefer that too, though unsure if it's appropriate to refactor it in this PR? I've just copied the existing code verbatim in this method. I guess I could add a commit with the refactoring, I agree it's nicer.

Reply:
I guess it should be okay to have this kind of refactor in this PR but I think we can wait for response from @emilio about this!

Reply:
Yeah, making code nicer is great over-all, and if done in separate commits it's even nicer :)

I personally prefer the first option (early-returning) better, but I'd be fine with both.